engine/launcher/src/main/scala/org/enso/launcher/upgrade/LauncherUpgrader.scala

```diff
@@ -217,28 +266,53 @@ class LauncherUpgrader(
 
   @scala.annotation.tailrec
   private def nextVersionToUpgradeTo(
-    release: LauncherRelease,
+    currentTargetRelease: LauncherRelease,
     availableVersions: Seq[SemVer]
   ): LauncherRelease = {
-    val recentEnoughVersions =
-      availableVersions.filter(
-        _.isGreaterThanOrEqual(release.minimumVersionToPerformUpgrade)
+    assert(
+      currentTargetRelease.minimumVersionToPerformUpgrade.isGreaterThan(
+        CurrentVersion.version
       )
+    )
+
+    // We look at older versions that are satisfying the minimum version
+    // required to upgrade to currentTargetRelease.
+    val recentEnoughVersions =
+      availableVersions.filter { possibleVersion =>
+        val canUpgradeToTarget = possibleVersion.isGreaterThanOrEqual(
+          currentTargetRelease.minimumVersionToPerformUpgrade
+        )
+        val isEarlierThanTarget =
+          possibleVersion.isLessThan(currentTargetRelease.version)
+        canUpgradeToTarget && isEarlierThanTarget
```

Comment:
Adding the second condition `isEarlierThanTarget` makes sure we 'progress' in going backwards in 'time'. Thus it breaks the infinite cycle - if we cannot find an earlier version that can upgrade to the current target, we will fail.